state_synchronizer/src/peer_manager.rs

```diff
@@ -70,16 +85,49 @@ impl PeerManager {
         self.get_active_upstream_peer_ids().is_empty()
     }
 
+    pub fn update_score(&mut self, peer_id: &PeerId, update_type: PeerScoreUpdateType) {
+        if let Some(peer_info) = self.peers.get_mut(peer_id) {
+            match update_type {
+                PeerScoreUpdateType::Success => {
+                    let new_score = peer_info.score + 1.0;
+                    peer_info.score = new_score.min(MAX_SCORE);
+                }
+                PeerScoreUpdateType::InvalidChunk => {
+                    let new_score = peer_info.score * 0.8;
+                    peer_info.score = new_score.max(MIN_SCORE);
+                }
+            }
+        }
+    }
+
     pub fn pick_peer(&mut self) -> Option<(PeerId, StateSynchronizerSender)> {
         let active_peers = self.get_active_upstream_peer_ids();
         debug!("[state sync] (pick_peer) state: {:?}", self.peers);
         if !active_peers.is_empty() {
-            let idx = thread_rng().gen_range(0, active_peers.len());
-            let peer_id = *active_peers[idx];
-            if let Some(sender) = self.get_network_sender(&peer_id) {
-                return Some((peer_id, sender));
-            } else {
-                debug!("[state sync] (pick_peer) no sender for {}", peer_id);
+            let weights: Vec<_> = active_peers
```

Comment:
looks verbose. Why not just replace get_active_upstream_peer_ids with get_active_peers that returns slice?